tests/tcp.rs

```diff
@@ -23,7 +23,103 @@ fn assert_error_kind<T>(res: io::Result<T>, kind: io::ErrorKind) {
 }
 
 async fn bind() -> std::result::Result<TcpListener, std::io::Error> {
-    TcpListener::bind((IpAddr::from(Ipv4Addr::UNSPECIFIED), PORT)).await
+    TcpListener::bind((Ipv4Addr::UNSPECIFIED, PORT)).await
+}
+
+#[test]
+fn connects_within_a_localhost() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.client("server", async {
+        let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, PORT)).await?;
+        tokio::spawn(async move {
+            loop {
+                let _ = listener.accept().await;
+            }
+        });
+
+        TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await?;
+
+        assert_error_kind(
+            TcpStream::connect((Ipv6Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        Ok(())
+    });
+
+    sim.client("serveripv6", async {
+        let listener = TcpListener::bind((Ipv6Addr::LOCALHOST, PORT)).await?;
+
+        tokio::spawn(async move {
+            loop {
+                let _ = listener.accept().await;
+            }
+        });
+
+        TcpStream::connect((Ipv6Addr::LOCALHOST, PORT)).await?;
+
+        assert_error_kind(
+            TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        Ok(())
+    });
+
+    sim.run()
+}
+
+#[test]
+fn sends_data_within_a_localhost() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.client("server", async {
+        let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, PORT)).await?;
+        tokio::spawn(async move {
+            if let Ok((mut stream, _)) = listener.accept().await {
+                let _ = stream.write_u8(1).await;
+                assert_eq!(2, stream.read_u8().await.unwrap());
+            }
+        });
+
+        let mut stream = TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await?;
+
+        assert_eq!(1, stream.read_u8().await?);
+        stream.write_u8(2).await?;
+        Ok(())
+    });
+
+    sim.run()
+}
+
+#[test]
+fn doesnt_allow_connection_outside_localhost() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.host("server", || async {
+        let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, PORT)).await?;
+
+        loop {
+            let _ = listener.accept().await;
+        }
+    });
+
+    sim.client("client", async {
+        assert_error_kind(
+            TcpStream::connect(("server", PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        assert_error_kind(
+            TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
```

Comment:
I'd avoid this case. It has nothing to do with the server's ip.

Reply:
I think it does, it makes sure if we bind something to `localhost` it's not available publicly.

Reply:
This is from a different host. How would it know which one to connect to even if localhost was exposed?